src/client/pkg/grpcutil/addr.go

```diff
@@ -0,0 +1,127 @@
+package grpcutil
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+const (
+	// DefaultPachdNodePort is the pachd kubernetes service's default
+	// NodePort.Port setting
+	DefaultPachdNodePort = 30650
+
+	// DefaultPachdPort is the pachd kubernetes service's default
+	// Port (often used with Pachyderm ELBs)
+	DefaultPachdPort = 650
+)
+
+var (
+	// ErrNoPachdAddress is returned by ParsePachdAddress when the input is an
+	// empty string
+	ErrNoPachdAddress = errors.New("no pachd address specified")
+	// DefaultPachdAddress is the default PachdAddress that should be used
+	// if none is otherwise specified. It's a loopback that should rely on
+	// port forwarding.
+	DefaultPachdAddress = PachdAddress{
+		Secured: false,
+		Host:    "0.0.0.0",
+		Port:    DefaultPachdNodePort,
+	}
+)
+
+// PachdAddress represents a parsed pachd address value
+type PachdAddress struct {
+	// Secured specifies whether grpcs should be used
+	Secured bool
+	// Host specifies the pachd address host without the port
+	Host string
+	// Port specifies the pachd port
+	Port uint16
+}
+
+// ParsePachdAddress parses a string into a pachd address, or returns an error
+// if it's invalid
+func ParsePachdAddress(value string) (PachdAddress, error) {
+	if value == "" {
+		return PachdAddress{}, ErrNoPachdAddress
+	}
+
+	secured := false
+
+	if strings.Contains(value, "://") {
+		// only parse the url if it contains a scheme, as net/url doesn't
+		// appropriately handle values without one
+
+		u, err := url.Parse(value)
+		if err != nil {
+			return PachdAddress{}, fmt.Errorf("could not parse pachd address: %v", err)
+		}
+
+		if u.Scheme != "grpc" && u.Scheme != "grpcs" {
+			return PachdAddress{}, fmt.Errorf("unrecognized scheme in pachd address: %s", u.Scheme)
+		}
+		if u.User != nil {
+			return PachdAddress{}, errors.New("pachd address should not include login credentials")
+		}
+		if u.RawQuery != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a query string")
+		}
+		if u.Fragment != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a fragment")
+		}
+		if u.Path != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a path")
+		}
+
+		value = u.Host
+		secured = u.Scheme == "grpcs"
+	}
+
+	// port always starts after last colon, but net.SplitHostPort returns an
+	// error on a hostport without a colon, which this might be
+	colonIdx := strings.LastIndexByte(value, ':')
+	host := value
+	port := uint16(DefaultPachdNodePort)
+	if colonIdx >= 0 {
+		maybePort, err := strconv.ParseUint(value[colonIdx+1:], 10, 16)
```

Comment:
would it make sense to call net.SplitHostPort here? Just because it seems like it might be shorter

Reply:
It doesn't work, see the comment above this block

Reply:
:sweat_smile:

Reply:
That relies on the URL always being parseable, but due to a separate issue, `url.Parse` doesn't handle values that don't include a protocol (which this is built to handle.)

Reply:
Ah, I'm suggesting adding a protocol at the beginning (the idea behind `if !strings.Contains(value, "://") { value = "grpc://" + value }`) precisely so that we *can* pass it to `url.Parse` in all cases, and then take advantage of its flexible port parsing, which handles cases like ipv6 and such.

Again, not trying to push this *too* hard, but I thought that adding the protocol might be easier than implementing a more flexible `net.SplitHostPort`, as the real implementation is fairly complex: https://golang.org/src/net/ipsock.go?s=4766:4832#L146

Reply:
Oh, yeah I see that in your code now. OK I can make that change.

Reply:
Done!